# Práctica 3 - Exploración de un dataset con pandas

## Objetivos

En esta práctica trabajarás con un dataset de clientes utilizando Python.

Al finalizar deberás ser capaz de:

- cargar datos desde un archivo CSV
- trabajar con un `DataFrame` de pandas
- identificar muestras y variables
- consultar la estructura de un dataset
- calcular estadísticas descriptivas
- seleccionar y filtrar datos
- analizar variables categóricas
- crear visualizaciones sencillas
- identificar las características \(X\) y la variable objetivo \(y\)
- determinar qué tipo de problema de Machine Learning podría resolverse

> **Importante:** no debes entrenar ningún modelo de Machine Learning en esta práctica.

## Contexto

Una empresa de telecomunicaciones quiere analizar el comportamiento de sus clientes. Dispone de información histórica y conoce qué clientes terminaron abandonando el servicio.

El objetivo futuro será desarrollar un sistema capaz de predecir:

> **¿Qué clientes tienen riesgo de abandonar la compañía?**

Para realizar la práctica utilizaremos el archivo `clientes.csv`.

| Variable | Descripción |
|---|---|
| `cliente_id` | Identificador del cliente |
| `edad` | Edad del cliente |
| `antiguedad` | Meses como cliente |
| `cuota_mensual` | Cuota mensual en euros |
| `consumo_datos` | Consumo mensual de datos en GB |
| `llamadas_soporte` | Número de llamadas a soporte |
| `incidencias` | Número de incidencias registradas |
| `tipo_contrato` | Mensual, anual o bienal |
| `satisfaccion` | Valoración del servicio de 1 a 5 |
| `abandono` | Indica si el cliente abandonó: Sí/No |

Cada fila corresponde a un cliente.

### Instrucciones

Completa las celdas de código marcadas con `# TODO`.

Cuando se solicite una interpretación, escribe tu respuesta en la celda Markdown correspondiente. No basta con ejecutar código: debes **interpretar los resultados**.

## Parte 1 - Preparación del entorno

### Ejercicio 1 - Importación

Importa NumPy, pandas y Matplotlib utilizando los alias convencionales.

In [2]:
# TODO: importa NumPy, pandas y matplotlib.pyplot
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Matplotlib is building the font cache; this may take a moment.


### Respuesta

Explica brevemente para qué utilizaremos cada biblioteca:

- **NumPy:** 
- **pandas:** 
- **Matplotlib:**

## Parte 2 - Carga e inspección inicial

### Ejercicio 2 - Cargar el dataset

Carga `clientes.csv` en un DataFrame denominado `df` y muestra sus primeras filas.

In [3]:
# TODO: carga el archivo clientes.csv en df
df = pd.read_csv('clientes.csv')
# TODO: muestra las primeras filas
df.head()

,cliente_id,edad,antiguedad,cuota_mensual,consumo_datos,llamadas_soporte,incidencias,tipo_contrato,satisfaccion,abandono
0,10001,23,61.0,32.16,11.9,1,2,Mensual,1,No
1,10002,62,79.0,45.71,17.4,2,0,Anual,2,No
2,10003,55,49.0,60.48,14.6,0,1,Mensual,5,No
3,10004,43,6.0,116.82,21.8,4,0,Anual,3,No
4,10005,43,70.0,36.13,8.8,2,2,Anual,1,Sí


### Respuesta

**¿Qué representa cada fila del DataFrame?**

Escribe aquí tu respuesta.

**¿Qué representa cada columna?**

Escribe aquí tu respuesta.

### Ejercicio 3 - Dimensiones

Obtén mediante código las dimensiones del dataset.

In [ ]:
# TODO: muestra las dimensiones del DataFrame

### Respuesta

- **Número de muestras:** 
- **Número de variables:**

### Ejercicio 4 - Columnas

Obtén mediante código los nombres de todas las columnas.

In [ ]:
# TODO: muestra los nombres de las columnas

### Respuesta

**¿Qué columna parece representar la variable que queremos predecir? ¿Por qué?**

Escribe aquí tu respuesta.

### Ejercicio 5 - Tipos de datos

Obtén los tipos de datos de todas las columnas.

In [ ]:
# TODO: consulta los tipos de datos

### Respuesta

**Variables numéricas:**

Escribe aquí tu respuesta.

**Variables categóricas:**

Escribe aquí tu respuesta.

**¿`cliente_id` debería considerarse una característica numérica simplemente porque contiene números? Razona la respuesta.**

Escribe aquí tu respuesta.

## Parte 3 - Estructura y calidad inicial

### Ejercicio 6 - Información general

Utiliza `info()` para analizar la estructura del dataset.

In [ ]:
# TODO: muestra la información general del DataFrame

### Respuesta

- ¿Cuántas entradas contiene?
- ¿Qué tipos de datos aparecen?
- ¿Todas las columnas contienen el mismo número de valores no nulos?
- ¿Parece existir algún valor ausente?
- ¿Cómo puedes deducirlo a partir de `info()`?

Escribe aquí tu análisis.

## Parte 4 - Estadística descriptiva

### Ejercicio 7 - Descripción estadística

Obtén las estadísticas descriptivas de las variables numéricas.

In [ ]:
# TODO: genera las estadísticas descriptivas

### Interpretación

A partir del resultado, indica:

- edad media;
- edad mínima;
- edad máxima;
- cuota mensual media;
- cuota mensual máxima;
- media de llamadas a soporte;
- media de incidencias;
- satisfacción media.

Escribe aquí una breve interpretación, no solo los valores.

### Ejercicio 8 - Cálculos individuales

Calcula mediante instrucciones independientes los valores solicitados.

In [ ]:
# TODO: media de edad


# TODO: mediana de edad


# TODO: cuota mínima


# TODO: cuota máxima


# TODO: número total de incidencias


# TODO: satisfacción media

### Comprobación

¿Coinciden estos resultados con los obtenidos mediante `describe()` cuando corresponde?

Escribe aquí tu respuesta.

## Parte 5 - Selección de información

### Ejercicio 9 - Selección de columnas

Crea un DataFrame `df_clientes` que contenga únicamente `edad`, `antiguedad`, `cuota_mensual`, `satisfaccion` y `abandono`. Muestra sus primeras cinco filas.

In [ ]:
# TODO: crea df_clientes


# TODO: muestra las primeras cinco filas

### Respuesta

**¿Sigue representando cada fila al mismo cliente que en el DataFrame original? Explica por qué.**

Escribe aquí tu respuesta.

## Parte 6 - Filtrado

### Ejercicio 10 - Clientes por edad

Obtén todos los clientes mayores de 50 años y calcula cuántos cumplen la condición.

In [ ]:
# TODO: filtra los clientes mayores de 50 años


# TODO: calcula cuántos son

### Ejercicio 11 - Clientes con incidencias

Selecciona los clientes que tengan `incidencias > 0` y determina cuántos son.

In [ ]:
# TODO: filtra los clientes con incidencias


# TODO: calcula cuántos son

### Ejercicio 12 - Condiciones combinadas

Obtén los clientes que cumplan simultáneamente:

\[
edad < 30
\]

y

\[
llamadas\_soporte \geq 3
\]

Muestra únicamente `cliente_id`, `edad`, `llamadas_soporte` y `abandono`.

In [ ]:
# TODO: realiza el filtro y muestra únicamente las columnas solicitadas

### Respuesta

**¿Qué operador has utilizado para combinar ambas condiciones?**

Escribe aquí tu respuesta.

### Ejercicio 13 - Segundo filtro combinado

Localiza clientes que cumplan:

\[
satisfaccion \leq 2
\]

y

\[
incidencias \geq 2
\]

Muestra las filas encontradas.

In [ ]:
# TODO: realiza el filtro

### Reflexión

**Observando únicamente estos resultados, ¿podemos afirmar que una satisfacción baja provoca el abandono?**

Justifica tu respuesta.

## Parte 7 - Variables categóricas

### Ejercicio 14 - Tipo de contrato

Obtén los diferentes valores de `tipo_contrato` y calcula cuántos clientes pertenecen a cada categoría.

In [ ]:
# TODO: muestra las categorías existentes


# TODO: calcula la frecuencia de cada categoría

### Respuesta

**¿Cuál es el tipo de contrato más frecuente?**

Escribe aquí tu respuesta.

### Ejercicio 15 - Variable `abandono`

Obtén las categorías existentes, el número de clientes de cada categoría y sus frecuencias relativas.

In [ ]:
# TODO: categorías existentes


# TODO: frecuencias absolutas


# TODO: frecuencias relativas

### Interpretación

Expresa las frecuencias relativas mediante porcentajes y redacta una breve conclusión sobre la distribución de `abandono`.

Escribe aquí tu respuesta.

## Parte 8 - Visualización

Todos los gráficos deben contener como mínimo:

- título;
- etiquetas de los ejes cuando correspondan;
- una representación clara.

### Ejercicio 16 - Distribución de edad

Construye un histograma de `edad`.

In [ ]:
# TODO: crea el histograma de edad

### Interpretación

**¿Qué información proporciona un histograma que no resulta tan fácil obtener observando directamente las filas del DataFrame?**

Escribe aquí tu respuesta.

### Ejercicio 17 - Abandono

Representa mediante un gráfico de barras el número de clientes que abandonaron y que no abandonaron.

In [ ]:
# TODO: crea el gráfico de barras

### Interpretación

Describe brevemente qué muestra el gráfico.

Escribe aquí tu respuesta.

### Ejercicio 18 - Cuota mensual

Crea un histograma para `cuota_mensual`.

In [ ]:
# TODO: crea el histograma de cuota_mensual

### Interpretación

Analiza visualmente la distribución.

Escribe aquí tu respuesta.

### Ejercicio 19 - Edad y cuota

Crea un diagrama de dispersión utilizando:

\[
x = edad
\]

\[
y = cuota\_mensual
\]

In [ ]:
# TODO: crea el diagrama de dispersión

### Interpretación

- ¿Qué representa cada punto?
- ¿Observas a simple vista alguna relación clara entre ambas variables?

No es necesario realizar un análisis estadístico.

Escribe aquí tu respuesta.

## Parte 9 - Identificación de $X$ e $y$

La empresa quiere construir en el futuro un modelo que determine si un cliente abandonará.

### Ejercicio 20 - Variable objetivo

Identifica y crea `y`. Después muestra sus cinco primeros valores.

In [ ]:
# TODO: crea y


# TODO: muestra sus cinco primeros valores

### Respuesta

**¿Por qué esta variable corresponde a \(y\)?**

Escribe aquí tu respuesta.

### Ejercicio 21 - Características

Crea `X` utilizando inicialmente estas características:

- `edad`
- `antiguedad`
- `cuota_mensual`
- `consumo_datos`
- `llamadas_soporte`
- `incidencias`
- `tipo_contrato`
- `satisfaccion`

Después comprueba sus dimensiones.

In [ ]:
# TODO: crea X


# TODO: muestra las dimensiones de X

### Respuesta

Si el dataset contiene \(n\) clientes, esperamos:

\[
X.shape = (n,\ ?)
\]

**¿Qué número debe aparecer en `?`? ¿Por qué?**

Escribe aquí tu respuesta.

## Parte 10 - Razonamiento de Machine Learning

### Ejercicio 22 - Identificación del problema

Queremos predecir:

`abandono → Sí / No`

Responde razonadamente:

1. ¿Estamos ante aprendizaje supervisado o no supervisado?
2. ¿Por qué?
3. ¿Es clasificación o regresión?
4. ¿Por qué?
5. ¿Es clasificación binaria o multiclase?
6. ¿Cuáles son las clases?

### Respuesta

Escribe aquí tu razonamiento.

## Parte 11 - Un cambio en el problema

Supongamos ahora que la empresa plantea:

> Queremos predecir cuál será el **consumo de datos del próximo mes** de cada cliente.

### Ejercicio 23

Responde:

1. ¿Cuál sería ahora la variable objetivo?
2. ¿Seguiríamos teniendo exactamente el mismo \(X\)?
3. ¿Sería clasificación o regresión?
4. ¿Seguiría siendo aprendizaje supervisado?
5. Justifica cada respuesta.

### Respuesta

Escribe aquí tu razonamiento.

## Parte 12 - Análisis crítico

### Ejercicio 24 - ¿Utilizarías `cliente_id`?

Los identificadores tienen valores como:

`10001, 10002, 10003, 10004, ...`

¿Utilizarías esta columna como característica para predecir el abandono?

No respondas únicamente «sí» o «no». Razona tu decisión.

### Respuesta

Escribe aquí tu respuesta.

### Ejercicio 25 - ¿Estamos preparados para entrenar?

Después del análisis anterior, alguien propone ejecutar directamente:

```python
modelo.fit(X, y)
```

¿Estamos preparados?

Identifica **al menos cuatro cuestiones** que deberíamos estudiar o resolver antes de entrenar el modelo.

No necesitas explicar todavía cómo resolverlas.

### Respuesta

1. 
2. 
3. 
4.

## Parte 13 - Conclusiones

### Ejercicio 26 - Informe final

Escribe entre **8 y 12 líneas** resumiendo lo que has descubierto.

Debes incluir como mínimo:

- tamaño del dataset
- principales tipos de variables
- comportamiento de `abandono`
- alguna observación de las estadísticas
- alguna observación de las visualizaciones
- cuáles son \(X\) e \(y\)
- tipo de problema de Machine Learning
- qué pasos deberían realizarse antes de entrenar

### Conclusiones

Escribe aquí tu informe.

---

# Ampliación opcional

Si terminas antes, investiga cómo obtener con pandas:

1. El número exacto de valores ausentes de cada columna.
2. El número de clientes agrupados por `tipo_contrato` y `abandono`.
3. La cuota mensual media de quienes abandonaron y de quienes no.
4. La satisfacción media de ambos grupos.

No utilices algoritmos de Machine Learning.

In [ ]:
# AMPLIACIÓN OPCIONAL

# 1. Valores ausentes por columna


# 2. Clientes agrupados por tipo_contrato y abandono


# 3. Cuota mensual media según abandono


# 4. Satisfacción media según abandono

# Antes de entregar

Comprueba que:

- has completado todas las celdas `TODO`
- has respondido las preguntas en Markdown
- todos los gráficos tienen título y etiquetas
- has reiniciado el entorno y ejecutado el notebook completo desde el principio
- no se producen errores
- no has entrenado ningún modelo

**Nombre recomendado del archivo de entrega:**

`Practica_03_Apellido_Nombre.ipynb`